util/src/test/scala/org/broadinstitute/dsde/rawls/util/RetrySpec.scala

```diff
@@ -0,0 +1,179 @@
+      testable.invocationCount should equal(4)
+      verify(testable.underlying, times(4)).info(anyString, any[Throwable])
+    }
+  }
+
+  it should "not retry upon success" in {
+    val testable = new TestRetry(system, setUpMockLogger)
+
+    val result = testable.retry()(() => testable.success)
+
+    // result should a success
+    // invocationCount should be 1
+    // should not have logged anything
+    whenReady(result) { res =>
+      res should be (42)
+      testable.invocationCount should equal(1)
+      verify(testable.underlying, never).info(anyString, any[Throwable])
+    }
+  }
+
+  it should "not retry if the predicate returns false" in {
+    val testable = new TestRetry(system, setUpMockLogger)
+
+    val result = testable.retry(_ => false)(() => testable.failure)
+
+    // result should be a failure
+    // invocationCount should be 1
+    // log count should be 1
+    whenReady(result.failed) { ex =>
+      ex shouldBe an [Exception]
+      ex should have message "test exception"
+      testable.invocationCount should equal(1)
+      verify(testable.underlying, times(1)).info(anyString, any[Throwable])
+    }
+  }
+
+  it should "log a custom error message" in {
+    val testable = new TestRetry(system, setUpMockLogger)
+    val customLogMsg = "custom"
+
+    val result = testable.retry(failureLogMessage = customLogMsg)(() => testable.failure)
+
+    // result should be a failure
+    // invocationCount should be 4 (1 initial run plus 3 retries)
+    // log count should be 4 (3 retries plus 1 failure message)
+    whenReady(result.failed) { ex =>
+      ex shouldBe an [Exception]
+      ex should have message "test exception"
+      testable.invocationCount should equal(4)
+      val argumentCaptor = captor[String]
+      verify(testable.underlying, times(4)).info(argumentCaptor.capture, any[Throwable])
+      argumentCaptor.getAllValues.asScala.foreach { msg =>
+        msg should startWith(customLogMsg)
+      }
+    }
+  }
+
+  it should "retry exponentially 6 times" in {
+    // Need to increase the patience config because exponential retries take longer
+    implicit val patienceConfig = PatienceConfig(timeout = scaled(Span(2, Minutes)))
+
+    val testable = new TestRetry(system, setUpMockLogger)
+
+    val result = testable.retryExponentially()(() => testable.failure)
+
+    // result should be a failure
+    // invocationCount should be 7 (1 initial run plus 6 retries)
+    // log count should be 7 (6 retries plus 1 failure message)
+    whenReady(result.failed) { ex =>
+      ex shouldBe an [Exception]
+      ex should have message "test exception"
+      testable.invocationCount should equal(7)
+      verify(testable.underlying, times(7)).info(anyString, any[Throwable])
+    }
+  }
+
+  it should "retry until a success" in {
+    val testable = new TestRetry(system, setUpMockLogger)
+
+    val result = testable.retryUntilSuccessOrTimeout()(100 milliseconds, 1 day)(() => testable.failureNTimes(10))
```

Comment:
Sure, 1 day is excessive. Though the idea here is to _not_ hit the timeout -- `whenReady` will give up anyway after 10 seconds.